### PR TITLE
Reformatting all grammars

### PR DESCRIPTION
**Pull request by mike-lischke**
This patch applies common formatting rules to all grammars in the repository. Naturally, it's very large - probably the largest patch you will ever get 😅 and if you don't like that, I can open multiple PRs. However, that will not lower the work of reviewing the changes (quite the opposite).

All grammars also got the used formatting rules attached, so if you want, you can do your own formatter run using [`antlr-format-cli`](https://www.npmjs.com/package/antlr-format-cli).

### CODE REVIEW

**Comment by mike-lischke** [comment]
Looks like there are 2-3 grammars that got errors. I'll investigate over the weekend.

**Comment by kaby76** [comment]
> Looks like there are 2-3 grammars that got errors. I'll investigate over the weekend.

antlr/antlr4 failed because the .errors needs to be remastered. The example .g4's were reformatted.

apt failed because there's a symbol conflict with "options" in the first rule -- it's a keyword. The Antlr4 tool does not seem to have a problem with "options=" but does with "options =". Bug in the Antlr4 Tool.

For sql/tsql, the error is again a symbol conflict. "options+=" vs "options +=". Bug in Antlr4 Tool.

**Comment by mike-lischke** [comment]
Ah thanks Ken, for the analysis. Shall I revert the formatting of the example .g4's or rather adjust what you call "the .errors" (no idea what that is)? For the other 2: we can just ignore them? I wonder why these symbol conflicts do not show up in other PRs, but maybe that is because these specific grammars are seldomly touched...

**Comment by kaby76** [comment]
> Ah thanks Ken, for the analysis. Shall I revert the formatting of the example .g4's or rather adjust what you call "the .errors" (no idea what that is)? For the other 2: we can just ignore them? I wonder why these symbol conflicts do not show up in other PRs, but maybe that is because these specific grammars are seldomly touched...

I guess just revert the formatting for the examples. The .errors files is the output from the parse. The example file itself has an error in it, and we're just checking that the parser works the same across all targets and hasn't changed from one version of the tool to the next.

The symbol conflict wasn't detected because "options=" gets through the Antlr4 tool, but "options =" does not. I think just rename the element label "options" to "options_" since the grammar should work regardless of formatting.

**Comment by mike-lischke** [comment]
I reverted the examples and fixed the symbol conflicts. However, I don't fully understand why `options` is actually a conflict. Is that target language specific? I saw the JS target fail, but don't remember options has a special meaning there.

**Comment by kaby76** [comment]
> I reverted the examples and fixed the symbol conflicts. However, I don't fully understand why `options` is actually a conflict. Is that target language specific? I saw the JS target fail, but don't remember options has a special meaning there.

It's not target specific. It's actually from the Antlr4 tool. It's kind of hard to see the error, but the build error is here: https://github.com/antlr/grammars-v4/actions/runs/6988622530/job/19016369650#step:22:1336. The Antlr4 tool just doesn't like the grammar with the space after the "options", whereas before reformatting there was no space after "options" and the tool seems to like the grammar.

**Comment by mike-lischke** [comment]
Simply jaw-dropping how long the validation takes in this repo... There are still a few failures, where I don't know what to do. It believe this patch is fine.

**Comment by kaby76** [comment]
> Simply jaw-dropping how long the validation takes in this repo... There are still a few failures, where I don't know what to do. It believe this patch is fine.

You changing every grammar more or less. Every grammar is tested for every target. And even though the work is split up on dozens of machines, a lot of the grammars--frankly--suck. That's because they contains terrible and numerous ambiguities and many fall back to full context. The testing is better nowadays because it only tests only those grammars what you changed. Before, it was up to 8 hours for any change. But, the other issue is that the MacOS servers on Github Actions are slow. So, it's possible what did work, no longer works because something in the target has changed.

So, here's a rundown on the failures.
* asm/ptx/ptx-isa-1.0 with Cpp: NAN is a symbol conflict for Cpp--apparently now for MacOS.
* ./icon ./lark .sieve ./smtlibv2 ./wavefront with Go target--symbol conflict apparently now with "start" as a parser rule.
* ./scala with Go target--timeout. This grammar is super slow. Go target should be removed from desc.xml to indicate it doesn't work.
* elixir with PHP target--stack overflow. The PHP has logical errors in the code. I tried to address some of the problems, but it was too much, and needed to work on other things. https://github.com/antlr/antlr-php-runtime/pull/34

I think it's fine to go. I'll clean up these grammars at some point.

**Comment by mike-lischke** [comment]
Hmm, probably some of the grammars should be removed if they don't work. It's great to have such an exhaustive test framework, but taking hours to run is really a pain. Anyway, we cannot fix all problems in a single PR.

@teverett any objections to merge this PR?

**Comment by teverett** [comment]
@mike-lischke  no objection to merging the PR.  I prefer not to remove any grammars however.  Thanks for this contribution.

I know the new QA tools @kaby76 has worked on are great, and have the ability to exclude grammar tests for certain combinations of grammar and language. Can we use those to exclude the grammars that are having trouble so that we have a clean build?

**Comment by mike-lischke** [comment]
Right, we should not do anything in addition to this already large patch. @kaby76 can you please review the patch? Thanks!

**Comment by kaby76** [comment]
> Right, we should not do anything in addition to this already large patch. @kaby76 can you please review the patch? Thanks!

It's fine. I created an issue to correct these grammars in a separate PR. It's best to get the bulk of the changes in before git conflicts start to become an issue.

**Comment by mike-lischke** [comment]
Is a formal approval required here to merge a patch (i.e. via the review process)? If not @teverett please proceed.

**Comment by mike-lischke** [comment]
That would require to add changes unrelated to the patch. As I understand it these errors existed before already, but some did not show until now ("options=") or have been ignored (?)(slow PHP and Go grammars, stack overflow). I assume fixing those issues will take quite a while and I cannot do that. Additionally, with every new patch you merge I have to touch this PR again to fix merge conflicts.

I would prefer to merge this PR, because it is not really the reason why some tests fail. But if someone who knows the nature of these problems better than me (@kaby76?) volunteers to fix them I'm ready to wait.

**Comment by kaby76** [comment]
I can make a PR asap to fix these problems.

**Comment by kaby76** [comment]
> Yes, these are problems that existed before this PR, #3847 I can make a PR asap to fix these problems. Then, you can merge the changes in to your PR.

The PR containing the fixes is https://github.com/antlr/grammars-v4/pull/3849

**Comment by mike-lischke** [comment]
This patch is simpler than I expected. Let me just cherry pick them into my PR here.

**Comment by kaby76** [comment]
Wow, the build is taking a while. Maybe I should fork a job for each `OS x target x grammar x build-script`, but I don't know what Github Actions would do with `3 OSes * 8 targets * 350 grammars * 2 build scripts + 1 static workflow for testing grammar issues = 16801` jobs being submitted all at once?!

**Comment by mike-lischke** [comment]
That's what I meant above :-D Maybe a [job matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs) helps?

**Comment by kaby76** [comment]
Sorry, I just realized I didn't set `cancel-in-progress`. That's a big reason why the build is so slow. https://github.com/antlr/grammars-v4/issues/3851

**Comment by mike-lischke** [comment]
¯\_(ツ)_/¯

**Comment by mike-lischke** [comment]
We are close now. Only PHP on Ubuntu failed. @kaby76 can you please take another look what's wrong this time?

**Comment by kaby76** [comment]
The difference between the last clean build of this PR ([here](https://github.com/antlr/grammars-v4/actions/runs/7004430153)) and the last build ([here](https://github.com/antlr/grammars-v4/actions/runs/7010383029)) is because we now have [my PR build change](https://github.com/antlr/grammars-v4/pull/3849) testing PHP on Ubuntu with Powershell. There's a bug I now see in the Powershell script template, so I cannot tell why [PHP failed](https://github.com/antlr/grammars-v4/actions/runs/7010383029/job/19070929695#step:22:18067) for grammars `./ada/ada2012 ./bicep ./c ./sql/hive/v4`. The grammars work for Ubuntu using the Bash script, so it's likely there is something wrong with the Powershell script or the environment. I just checked each of these on my Ubuntu machine, and they all pass.

I suggest we merge it in ASAP so that additional merges cause more issues.

**Comment by kaby76** [comment]
Oh geezus H.... Sorry, my bug.

The environment is not setting up PHP for Powershell. So, I have no idea what PHP it *is* running. Works fine in Bash, where [it does this step](https://github.com/antlr/grammars-v4/blob/c1e4b59f00501c94b310657e954f620aa3cf4a0e/.github/workflows/main.yml#L165-L174).

**Comment by mike-lischke** [comment]
So, shall we wait for your fix for PHP or still merge?

**Comment by kaby76** [comment]
> So, shall we wait for your fix for PHP or still merge?

Let's get your PR in ASAP because it's huge, and may get whacked by another merge. I'll clean up the workflow issue afterwards. The grammars work fine for PHP.

**Comment by kaby76** [comment]
So, I fixed [this problem](https://github.com/antlr/grammars-v4/pull/3857). @mike-lischke If you want to rebase, that's up to you. Or, @teverett could merge my PR in first, then merge PR #3843 (there shouldn't be any merge conflicts) and it should all build and test fine. Or, just merge #3843 first, ignore the error with PHP/Ubuntu, and then PR #3857,

**Comment by mike-lischke** [comment]
I'd prefer to merge my patch first.

**Comment by teverett** [comment]
Well, there is a merge conflict now with 

sql/mysql/Positive-Technologies/MySqlParser.g4

**Comment by mike-lischke** [comment]
@harveyyue committed something, even though we said we don't want any commit until this patch is in. How can that be? How can he push something without a review process? 

@teverett What if someone else decides to commit while we are waiting for the next validation and then the next and the next? Where should that end?

**Comment by kaby76** [comment]
> @harveyyue committed something, even though we said we don't want any commit until this patch is in. How can that be? How can he push something without a review process?
> 
> @teverett What if someone else decides to commit while we are waiting for the next validation and then the next and the next? Where should that end? I assumed you are the owner of this repo and have the last word what goes in 🤔

@KvanTTT merged PR https://github.com/antlr/grammars-v4/pull/3858, and doesn't seem to be in the loop on this PR. Hopefully, no more merges into this repo until this PR is merged first.

**Comment by mike-lischke** [comment]
OK, my apologies for the wrong accusation. It wasn't clear to me from the patch that this was a merged PR.

**Comment by mike-lischke** [comment]
In addition to the already mentioned PHP on Ubuntu problem now the Dart on macOS timed out. How to proceed @teverett? Maybe you can restart the task and maybe this time it does not time out. Or just merge straight away?

**Comment by kaby76** [comment]
The dart2 problem is a network issue. Github actions are pounding on servers for dart libraries and they cannot keep up with the requests. Sometimes builds work sometimes not because of load. We've seen this many times before. Merge please.

**Comment by teverett** [comment]
@kaby76  @mike-lischke  thanks so much for this.

File: sql/snowflake/SnowflakeParser.g4 [hunk 1]
```diff
@@ -256,63 +245,83 @@ execute_task
...
 grant_ownership
-    : GRANT OWNERSHIP
-         ( ON ( object_type_name object_name | ALL object_type_plural IN ( DATABASE id_ | SCHEMA schema_name ) )
-         | ON FUTURE object_type_plural IN ( DATABASE id_ | SCHEMA schema_name )
-         )
-         TO ROLE id_ ( ( REVOKE | COPY ) CURRENT GRANTS )?
+    : GRANT OWNERSHIP (
+        ON (
+            object_type_name object_name
+            | ALL object_type_plural IN ( DATABASE id_ | SCHEMA schema_name)
```
Comment (mlorek) [inline sql/snowflake/SnowflakeParser.g4:263]: @mike-lischke any reason why the space characters after `( DATABASE` were kept?

File: sql/snowflake/SnowflakeParser.g4 [hunk 2]
```diff
@@ -256,63 +245,83 @@ execute_task
...
+            | INTEGRATION
+        ) object_name
+        | (schema_privileges | ALL PRIVILEGES?) ON (
+            SCHEMA schema_name
+            | ALL SCHEMAS IN DATABASE id_
+        )
+        | ( schema_privileges | ALL PRIVILEGES?) ON ( FUTURE SCHEMAS IN DATABASE id_)
+        | (schema_object_privileges | ALL PRIVILEGES?) ON (
+            object_type object_name
+            | ALL object_type_plural IN ( DATABASE id_ | SCHEMA schema_name)
```
Comment (mlorek) [inline sql/snowflake/SnowflakeParser.g4:286]: same here `(`

File: sql/snowflake/SnowflakeParser.g4 (hunk 1 above)
Comment (mike-lischke) [reply sql/snowflake/SnowflakeParser.g4:263]: Well, no specific reason actually. Might be a formatter bug. I believe that's coming up when an alt is placed entirely on one line and contains groups.

Comment (mlorek) [reply sql/snowflake/SnowflakeParser.g4:263]: sry, space character after `(` and before `DATABASE`

**Pull request merged**